### PR TITLE
Propagate user timeout limits to Redis [LIMITS]

### PR DESCRIPTION
**Pull request by jgoizueta**
We need to replicate the user_timeout and database_timeout to redis, as well as new per-user parameters for total rendering timeout, for authorized and public access.

### CODE REVIEW

**Comment by jgoizueta** [comment]
We now have 4 parameters stored in the users metadata table; These parameters can be synced from Central:

* Timeout for authenticated db accesses: `user_timeout` (milliseconds): default value is 300000; It is propagated to
  Redis as `limits:timeout:{username}.db`. It is applied to the user's database role as default statement timeout.
* Timeout for public db accesses: `database_timeout` (milliseconds): default value is 300000; It is propagated to
  Redis as `limits:timeout:{username}.db_public`. It is applied to the user's database as default statement timeout.
* Timeout for authenticated tile rendering: `user_render_timeout` (milliseconds): default is NULL; It is propagated to
  Redis as `limits:timeout:{username}.render`. Intended for comsuption at the tiler.
  and defaults it to the tiler configuration parameter `renderer.mapnik.limits.render`.
* Timeout for non-authenticated tile rendering: `database_render_timeout` (milliseconds): default is NULL; It is propagated to
  Redis as `limits:timeout:{username}.render_public`.  Intended for comsuption at the tiler.

The first two parameters (sql) already existed in the metadata database (and in Central) and were applied to the dbs, but not stored in redis.

Note that for the last two parameters (render) the tiler will apply the next order of precedence:
1. Existing redis `limits:tiler:{username}.render` (in use only for a few users which had some really huge polygons)
2. New redis `limits:timeout:{username}.render` or `.render_public`
3. Existing `renderer.mapnik.limits.render` in the tiler static configuration.

**Comment by jgoizueta** [comment]
@juanignaciosl : this adds two new user columns user_render_timeout and database_render_timeout  which will be synced from central (to be added in central plans and users too) and which are propagated to redis (under a new key, also with the existing columns user_timeout and database_timeout).

File: app/models/user.rb [hunk 1]
```diff
@@ -911,33 +914,42 @@ def key
...
+                          'soft_obs_general_limit',    soft_obs_general_limit,
+                          'mapzen_routing_quota',      mapzen_routing_quota,
+                          'soft_mapzen_routing_limit', soft_mapzen_routing_limit,
+                          'google_maps_client_id',     google_maps_key,
+                          'google_maps_api_key',       google_maps_private_key,
+                          'period_end_date',           period_end_date,
+                          'geocoder_provider',         geocoder_provider,
+                          'isolines_provider',         isolines_provider,
+                          'routing_provider',          routing_provider
+    $users_metadata.HMSET timeout_key,
```
Comment (juanignaciosl) [inline app/models/user.rb]: A rake task to store timeouts for current user is needed, isn't it?

File: db/migrate/20170704094757_add_render_timeouts_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
```
Comment (juanignaciosl) [inline db/migrate/20170704094757_add_render_timeouts_to_users.rb]: Don't forget to move this to a different PR (I set "do not merge yet" label because of this).

File: db/migrate/20170704094757_add_render_timeouts_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  Proc.new do
+    add_column :users, :user_render_timeout, :integer, null: true
+    add_column :users, :database_render_timeout, :integer, null: true
```
Comment (juanignaciosl) [inline db/migrate/20170704094757_add_render_timeouts_to_users.rb]: Don't we want defaults? If possible, defaults + not null makes everything easier.

File: app/models/user.rb (hunk 1 above)
Comment (jgoizueta) [inline app/models/user.rb]: Everything should keep working as now with the new timeouts missing in Redis: the consumer (the Tiler/SQL API) will have proper default values when the changes to use these limits are deployed. 
On the other hand, Central rake tasks are being prepared to change tiemeouts from there.

File: db/migrate/20170704094757_add_render_timeouts_to_users.rb (hunk 3 above)
Comment (jgoizueta) [inline db/migrate/20170704094757_add_render_timeouts_to_users.rb]: We intended to use NULL as the default (the ongoing changes in the tiler replace any empty/missing value in Redis by the defaults configured in the tiler).

Are there any good reasons to prefer a non-NULL default? (we could use 0 in that case and use the same convention in the tiler)

Comment (juanignaciosl) [inline db/migrate/20170704094757_add_render_timeouts_to_users.rb]: In general, if you can assume that there's a value, everything's easier and less error-prone. But keep `null` if you prefer.

File: app/models/user.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/models/user.rb]: Ok!
